File: pytest-intent/pytest_intent/meta_tests.py
"""Reusable meta-test functions for pytest-intent projects.

Drop-in usage in any project's `tests/test_meta.py`:

    from pytest_intent.meta_tests import test_intent_schema, test_intent_coverage  # noqa: F401

Both meta-tests use override-able fixtures so projects can layout intent specs and
tests across multiple files / directories. Defaults assume the simple case:
  - intent.yaml at pack_root
  - tests/ at pack_root

Override paths via fixtures in your conftest.py for multi-file / multi-dir layouts:
    @pytest.fixture
    def intent_yaml_paths(pack_root):
        return [pack_root / "intent" / "users.yaml", pack_root / "intent" / "permissions.yaml"]

    @pytest.fixture
    def intent_tests_dirs(pack_root):
        return [pack_root / "src" / "users" / "tests", pack_root / "src" / "auth" / "tests"]
"""

from __future__ import annotations

from pathlib import Path

import pytest

from .coverage import collect_annotated_tests, coverage_violations
from .schema import check_schema, parse_intent_yaml


@pytest.fixture
def intent_yaml_paths(pack_root: Path) -> list[Path]:
    """Default: single intent.yaml at pack_root. Override for multi-file layouts."""
    return [pack_root / "intent.yaml"]


@pytest.fixture
def intent_tests_dirs(pack_root: Path) -> list[Path]:
    """Default: single tests/ dir at pack_root. Override for multi-dir layouts."""
    return [pack_root / "tests"]


def _merged_claims(paths: list[Path]) -> dict[str, dict[str, object]]:
    """Union claims across all intent.yaml sources. On key collision, later wins."""
    out: dict[str, dict[str, object]] = {}
    for p in paths:
        if not p.exists():
            continue
        out.update(parse_intent_yaml(p))
    return out


def _merged_annotated(dirs: list[Path]) -> dict[str, list[str]]:
    """Union @intent-annotated tests across all test directories."""
    out: dict[str, list[str]] = {}
    for d in dirs:
        for cid, refs in collect_annotated_tests(d).items():
            out.setdefault(cid, []).extend(refs)
    return out


def test_intent_schema(intent_yaml_paths: list[Path]) -> None:
    """Every claim across all intent.yaml sources conforms to schema."""
    claims = _merged_claims(intent_yaml_paths)
    if not claims:
        return  # nothing declared yet
    violations = check_schema(claims)
    assert not violations, "intent.yaml schema violations:\n" + "\n".join("  " + v for v in violations)


def test_intent_coverage(intent_yaml_paths: list[Path], intent_tests_dirs: list[Path]) -> None:
    """Every claim has ≥1 @intent-decorated test; every annotated test references a real claim."""
    claims = _merged_claims(intent_yaml_paths)
    if not claims:
        return
    annotated = _merged_annotated(intent_tests_dirs)
    violations = coverage_violations(claims, annotated)
    assert not violations, "intent ↔ test coverage violations:\n" + "\n".join("  " + v for v in violations)

File: vitest-intent/src/meta-tests.ts
/**
 * Reusable meta-tests for vitest-intent projects.
 *
 * Drop-in usage in any project's `tests/meta.test.ts`:
 *
 *     import { registerIntentMetaTests } from 'vitest-intent/meta-tests';
 *     registerIntentMetaTests({ packRoot: __dirname + '/..' });
 *
 * Or for multi-file / multi-dir layouts:
 *
 *     registerIntentMetaTests({
 *       intentYamlPaths: [path.join(root, 'intent', 'a.yaml'), path.join(root, 'intent', 'b.yaml')],
 *       testsDirs: [path.join(root, 'src', 'a', 'tests'), path.join(root, 'src', 'b', 'tests')],
 *     });
 */

import { existsSync } from "node:fs";
import { join } from "node:path";

import { test, expect } from "vitest";

import { collectAnnotatedTests, coverageViolations, type AnnotatedTests } from "./coverage.js";
import { checkSchema, parseIntentYaml, type Claims } from "./schema.js";

export interface RegisterOptions {
  /** Project root containing intent.yaml + tests/, used as a default for both lists. */
  packRoot?: string;
  /** Override: explicit list of intent.yaml file paths. */
  intentYamlPaths?: string[];
  /** Override: explicit list of test directories to scan. */
  testsDirs?: string[];
}

/**
 * Registers two meta-tests with vitest:
 *   - intent.schema  — every claim conforms to required fields + valid values
 *   - intent.coverage — every claim has a test, every annotated test has a claim
 *
 * Both tests pass vacuously if no claims are declared (empty / missing intent.yaml).
 */
export function registerIntentMetaTests(opts: RegisterOptions = {}): void {
  const intentPaths = resolveIntentPaths(opts);
  const testDirs = resolveTestDirs(opts);

  test("intent.schema — every claim conforms to schema", () => {
    const claims = mergedClaims(intentPaths);
    if (Object.keys(claims).length === 0) return;
    const violations = checkSchema(claims);
    expect(violations, violations.join("\n  ")).toEqual([]);
  });

  test("intent.coverage — claims and tests are 1:1", () => {
    const claims = mergedClaims(intentPaths);
    if (Object.keys(claims).length === 0) return;
    const annotated = mergedAnnotated(testDirs);
    const violations = coverageViolations(claims, annotated);
    expect(violations, violations.join("\n  ")).toEqual([]);
  });
}

function resolveIntentPaths(opts: RegisterOptions): string[] {
  if (opts.intentYamlPaths && opts.intentYamlPaths.length > 0) {
    return opts.intentYamlPaths;
  }
  if (opts.packRoot) {
    return [join(opts.packRoot, "intent.yaml")];
  }
  return ["intent.yaml"];
}

function resolveTestDirs(opts: RegisterOptions): string[] {
  if (opts.testsDirs && opts.testsDirs.length > 0) {
    return opts.testsDirs;
  }
  if (opts.packRoot) {
    return [join(opts.packRoot, "tests")];
  }
  return ["tests"];
}

function mergedClaims(paths: string[]): Claims {
  const out: Claims = {};
  for (const p of paths) {
    if (!existsSync(p)) continue;
    Object.assign(out, parseIntentYaml(p));
  }
  return out;
}

function mergedAnnotated(dirs: string[]): AnnotatedTests {
  const out: AnnotatedTests = {};
  for (const d of dirs) {
    const found = collectAnnotatedTests(d);
    for (const [cid, refs] of Object.entries(found)) {
      if (!out[cid]) out[cid] = [];
      out[cid].push(...refs);
    }
  }
  return out;
}
